toolchain/lexer/tokenized_buffer.cpp

```diff
@@ -907,33 +905,26 @@ auto TokenizedBuffer::SourceBufferLocationTranslator::GetLocation(
   int line_number = line_it - buffer_->line_infos_.begin();
   int column_number = offset - line_it->start;
 
-  llvm::StringRef line;
-
-  // We might still be lexing the last line. If so, check to see if there are
-  // any newline characters between the position we've finished lexing up to
-  // and the given location.
-  if (incomplete_line_info && column_number > *last_line_lexed_to_column_) {
-    column_number = *last_line_lexed_to_column_;
-    int64_t start = line_it->start;
-    for (int64_t i = line_it->start + *last_line_lexed_to_column_; i != offset;
-         ++i) {
-      if (buffer_->source_->text()[i] == '\n') {
-        start = i;
-        ++line_number;
-        column_number = 0;
-      } else {
-        ++column_number;
-      }
+  // Start by grabbing the buffer. If the line isn't fully lexed, the length
+  // will be npos and the buffer will be grabbed from the start; we'll then
+  // adjust the length.
+  llvm::StringRef line =
+      buffer_->source_->text().substr(line_it->start, line_it->length);
+  if (line_it->length == static_cast<int32_t>(llvm::StringRef::npos)) {
+    CARBON_CHECK(column_number < *last_line_lexed_to_column_ ||
+                 line.substr(*last_line_lexed_to_column_,
+                             column_number - *last_line_lexed_to_column_)
+                         .count('\n') == 0)
```

Comment:
I think there can be newlines between `*last_line_lexed_to_column_` and `column_number`, eg in:

```
var s: String = '''
  error here: '''
```

... our last line will be line 1, and it'll be lexed to the start of the `'''`, but the error column will be on a subsequent line. But there can't be newlines in the initial `*last_line_lexed_to_column_` portion (because if there were, we'd have created new lines for them already as part of lexing to that column).

Reply:
Hm. I see, we get away with this because the only place we lex a multi-line token (or more generally, do any lookahead past a newline) is in `LexStringLiteral`, and it happens to produce no errors that point into the source buffer until *after* it's scanned the literal and called `HandleNewline` for each newline in it.

This seems fragile, but I suppose the alternative is to add code here that we can't test to handle a case that doesn't currently happen.